File: backend/src/main/java/at/ac/tuwien/sepr/groupphase/backend/service/ArtistService.java
package at.ac.tuwien.sepr.groupphase.backend.service;

import at.ac.tuwien.sepr.groupphase.backend.endpoint.dto.artist.ArtistCreateDto;
import at.ac.tuwien.sepr.groupphase.backend.endpoint.dto.artist.DetailedArtistDto;
import at.ac.tuwien.sepr.groupphase.backend.endpoint.dto.artist.SimpleArtistDto;
import at.ac.tuwien.sepr.groupphase.backend.entity.Artist;

import java.util.List;

/**
 * Service for managing artists.
 */
public interface ArtistService {

    /**
     * Find all artists.
     *
     * @return list of all artists
     */
    List<Artist> findAll();

    /**
     * Find artist by ID.
     *
     * @param id the artist ID
     * @return the artist entity
     * @throws at.ac.tuwien.sepr.groupphase.backend.exception.NotFoundException if artist not found
     */
    Artist findById(Long id);

    /**
     * Create a new artist.
     *
     * @param artist the artist to create
     * @return the created artist
     * @throws at.ac.tuwien.sepr.groupphase.backend.exception.ValidationException if artist data is invalid
     */
    Artist create(Artist artist);

    /**
     * Search artists by name (case-insensitive, partial match).
     *
     * @param name the name to search for
     * @return list of matching artists
     */
    List<Artist> searchByName(String name);

    /**
     * Search artists by name and include bands where the artist is a member.
     * This returns both direct matches and any bands containing the artist.
     *
     * @param name the name to search for
     * @return list of matching artists and their bands
     */
    List<Artist> searchArtistsWithBands(String name);

    /**
     * Create artist from DTO including member resolution for bands.
     * If the artist is a band, member IDs are resolved to actual artist entities.
     *
     * @param dto the artist creation data
     * @return the created artist as DTO
     * @throws at.ac.tuwien.sepr.groupphase.backend.exception.ValidationException if data is invalid
     * @throws at.ac.tuwien.sepr.groupphase.backend.exception.NotFoundException   if member artist not found
     */
    SimpleArtistDto createFromDto(ArtistCreateDto dto);

    /**
     * Search artists by name and optionally include bands.
     * Returns DTOs with all mapping already applied.
     *
     * @param name         the name to search for
     * @param includeBands if true, includes bands where artist is a member
     * @return list of matching artists as DTOs
     */
    List<SimpleArtistDto> searchArtistsAsDto(String name, boolean includeBands);

    /**
     * Get all artists as DTOs.
     *
     * @return list of all artists as DTOs
     */
    List<SimpleArtistDto> findAllAsDto();

    /**
     * Get artist by ID as detailed DTO including band members.
     *
     * @param id the artist ID
     * @return the artist as detailed DTO
     * @throws at.ac.tuwien.sepr.groupphase.backend.exception.NotFoundException if artist not found
     */
    DetailedArtistDto findByIdAsDto(Long id);
}

File: frontend/src/app/services/artist.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../environments/environment';
import { Artist } from '../dtos/artist';

/**
 * Service for managing artist-related operations.
 * Provides functionality for artist creation, search, retrieval, and querying events by artist.
 * Supports both individual artists and bands.
 * Handles communication with the backend artists API endpoint.
 */
@Injectable({
  providedIn: 'root'
})
export class ArtistService {
  private baseUri: string = environment.backendUrl + '/artists';

  constructor(private http: HttpClient) {}

  /**
   * Retrieves all artists and bands from the system.
   *
   * @returns Observable of all artists/bands as an array
   */
  getAllArtists(): Observable<Artist[]> {
    return this.http.get<Artist[]>(this.baseUri);
  }

  /**
   * Retrieves detailed information for a specific artist by their ID.
   * Includes information about band memberships if applicable.
   *
   * @param id - The unique identifier of the artist
   * @returns Observable of the artist details
   * @throws NotFoundException if the artist with the given ID does not exist
   */
  getArtistById(id: number): Observable<Artist> {
    return this.http.get<Artist>(`${this.baseUri}/${id}`);
  }

  /**
   * Creates a new artist or band in the system.
   *
   * @param artist - The artist data to create
   * @param artist.name - The name of the artist or band
   * @param artist.isBand - Optional flag indicating if this is a band (default: false)
   * @param artist.memberIds - Optional array of artist IDs who are members (only for bands)
   * @returns Observable of the created artist with assigned ID
   * @throws ValidationException if the artist data is invalid
   * @throws NotFoundException if referenced member IDs do not exist
   */
  createArtist(artist: { name: string; isBand?: boolean; memberIds?: number[] }): Observable<Artist> {
    return this.http.post<Artist>(this.baseUri, artist);
  }

  /**
   * Searches for artists by name with optional band inclusion.
   * Supports searching for both individual artists and band names.
   *
   * @param name - The search term for artist/band name (supports partial matching)
   * @param includeBands - If true, also returns bands where the artist is a member (default: true)
   * @returns Observable of artists matching the search criteria
   */
  searchArtists(name: string, includeBands: boolean = true): Observable<Artist[]> {
    let params = new HttpParams()
      .set('name', name)
      .set('includeBands', includeBands.toString());
    return this.http.get<Artist[]>(`${this.baseUri}/search`, { params });
  }

  /**
   * Retrieves all events featuring a specific artist or band.
   * Useful for displaying artist-specific event listings.
   *
   * @param artistId - The unique identifier of the artist or band
   * @returns Observable of all events featuring the specified artist
   * @throws NotFoundException if the artist does not exist
   */
  getEventsByArtist(artistId: number): Observable<any[]> {
    return this.http.get<any[]>(`${this.baseUri}/${artistId}/events`);
  }
}
